src/controls/DragControls.ts

```diff
@@ -0,0 +1,266 @@
+import { Camera, EventDispatcher, Intersection, Matrix4, Object3D, Plane, Raycaster, Vector2, Vector3 } from 'three'
+
+class DragControls extends EventDispatcher {
+  enabled: boolean
+  transformGroup: boolean
+
+  private _objects: Object3D[]
+  private _camera: Camera
+  private _domElement: HTMLElement
+
+  private _plane: Plane
+  private _raycaster: Raycaster
+
+  private _mouse: Vector2
+  private _offset: Vector3
+  private _intersection: Vector3
+  private _worldPosition: Vector3
+  private _inverseMatrix: Matrix4
+  private _intersections: Intersection[]
+
+  private _selected: Object3D | null
+  private _hovered: Object3D | null
+
+  constructor(_objects: Object3D[], _camera: Camera, _domElement: HTMLElement) {
+    super()
+
+    this._objects = _objects
+    this._camera = _camera
+    this._domElement = _domElement
+
+    this._plane = new Plane()
+    this._raycaster = new Raycaster()
+
+    this._mouse = new Vector2()
+    this._offset = new Vector3()
+    this._intersection = new Vector3()
+    this._worldPosition = new Vector3()
+    this._inverseMatrix = new Matrix4()
+    this._intersections = []
+    this._selected = null
+    this._hovered = null
+
+    this.activate()
+
+    // API
+
+    this.enabled = true
+    this.transformGroup = false
+  }
+
+  activate = () => {
+    this._domElement.addEventListener('pointermove', this.onPointerMove)
+    this._domElement.addEventListener('pointerdown', this.onPointerDown)
+    this._domElement.addEventListener('pointerup', this.onPointerCancel)
+    this._domElement.addEventListener('pointerleave', this.onPointerCancel)
+    this._domElement.addEventListener('touchmove', this.onTouchMove)
+    this._domElement.addEventListener('touchstart', this.onTouchStart)
+    this._domElement.addEventListener('touchend', this.onTouchEnd)
+  }
+
+  deactivate = () => {
+    this._domElement.removeEventListener('pointermove', this.onPointerMove)
+    this._domElement.removeEventListener('pointerdown', this.onPointerDown)
+    this._domElement.removeEventListener('pointerup', this.onPointerCancel)
+    this._domElement.removeEventListener('pointerleave', this.onPointerCancel)
+    this._domElement.removeEventListener('touchmove', this.onTouchMove)
+    this._domElement.removeEventListener('touchstart', this.onTouchStart)
+    this._domElement.removeEventListener('touchend', this.onTouchEnd)
+
+    this._domElement.style.cursor = ''
+  }
+
+  // TODO: confirm if this can be removed?
+  dispose = () => this.deactivate()
+
+  getObjects = () => this._objects
+
+  onMouseMove = (event: MouseEvent) => {
```

Comment:
Mmm in the three version they are "public", should we restrict them?

Reply:
I think we should because they manipulate the internal state of the controller, not that anyone with a brain would overwrite them, it's probably safer.

Reply:
So we should do this in every controller maybe